File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [-1] * n
        st = []
        for i in range(2 * n):
            idx = i % n
            while st and nums[st[-1]] < nums[idx]:
                ans[st[-1]] = nums[idx]
                st.pop()
            if i < n:
                st.append(idx)

        return ans 

File: LeetCode/0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n, -1);
        stack<int> st; 
        for (int i = 0; i < 2 * n; i++) {
            int idx = i % n;
            
            while (!st.empty() && nums[st.top()] < nums[idx]) {
                ans[st.top()] = nums[idx];
                st.pop();
            }
            
            if (i < n) {
                st.push(idx);
            }
        }
        
        return ans;
    }
};